### ISSUE TITLE
read.ctd() should have a maximum limit for number of items with the same name

### ISSUE THREAD

**Issue by clayton33** [dankelley/oce#1686]
Brought this up during a virtual f2f, I've been running into this issue where there are too many items in the header of some `.ODF` files. I've found that for these files with very long header's, it take a while to read them in as well. Wondering if it's worth setting a maximum limit. 

```
Error in read.odf(file = file, columns = columns, debug = debug - 1) : 
  cannot have more than 10000 items of the same name in ODF metadata; rerun with debug=5 to diagnose
```

I will send along a few files where I've encountered this error sometime in the next few hours.

**Comment by dankelley** [comment]
Just for the record, so @richardsc will see the issue, the problem data files have a **lot** of lines lines appear to be generated automatically, and are of the form
```
  PROCESS='Nulled the CRAT_01 value "0.79411" corresponding to the pressure value "701" in record "700"',
  PROCESS='Nulled the CRAT_01 value "0.7939" corresponding to the pressure value "702" in record "701"',
...
  PROCESS='Nulled the CRAT_02 value "0.79213" corresponding to the pressure value "729" in record "728"',
  PROCESS='Nulled the ALTB_01 value "-3.13" corresponding to the pressure value "2" in record "1"',
  PROCESS='Nulled the ALTB_01 value "40.69" corresponding to the pressure value "3" in record "2"',
 ... 
```
and so forth, over depth ranges and suites of variables.

**Comment by dankelley** [comment]
The problem is that `read.odf` is trying to decode all of those lines, and to store them in a big list called `PROCESS`, for inclusion in the data object.

This processing is slow, and so reading can be impracticably slow for large files, e.g.  `CTD_HUD2013037_042_1_DN.ODF` file, which has 13080 lines.

It's possible I can find a way to speed up the code, which is at https://github.com/dankelley/oce/blob/58c0faeeab7ef3261015804e390ce25f151e089c/R/odf.R#L1044 so that it might parse these lines faster, and I'll take a look at that.  (I had no idea that we could get more than maybe 5 repeats, so I would not have been thinking of speed when I wrote that part of the code.)

But even if I can halve the time, it would be way too long.  And, besides, do we really want over 10,000 entries in the metadata of an object?  I can't imagine wanting to use that for anything.

A crude solution comes to mind: I could add an argument called maybe `exclude`, which would be a regexp for lines to ignore in the file.  Calling it with
```R
read.odf(file, exclude="^[ ]*PROCESS='Nulled the.*$")
```
would be something that I would certainly use, if faced with such a file.  I'm fairly certain this would be fast, because `grep` is fast.  This would be set to `exclude=NULL`, meaning not to exclude any files.  And I would propose to change this message from 10,000 lines to maybe 500 lines, at which point I'd spit out a warning saying to consider using `exclude=`, if the speed is too slow.

Q for @clayton33 and @richardsc do you object to this `exclude` idea?  I know it can be made to work easily, whereas trying to find a way to speed up the processing might be tricky.

**Comment by clayton33** [comment]
In favour of  the `exclude` idea.

**Comment by dankelley** [comment]
I am adding the following to the `R/odf.R`, in the Roxygen documentation
code for `read.odf()`.
```
#' @param headerExclude either a character value holding a regular
#' expression that is used with [grep()] to remove lines from the
#' header before processing, or `NULL` (the default), meaning
#' not to exclude any such lines.  The purpose of this argument
#' is to solve problems with some files, which can have
#' thousands of lines that indicate details that are may be of
#' little value in processing.  For example, some files have thousands
#' of lines that would be excluded by using
#' `headerExclude="PROCESS='Nulled the .* value"` in the funcion call.
```

**Comment by dankelley** [comment]
I'm looking at the code. The `exclude` argument should speed things up, e.g.
```R
system.time(d <- read.oce("~/Desktop/CTD_HUD2013037_042_1_DN.ODF",headerExclude="PROCESS='Nulled the"))
   user  system elapsed 
  2.480   0.041   2.541 
```

This code is not yet pushed to GH.

Also, I see why it's so slow to process those lines.  It has to do with developing variable names that don't conflict with known names.  Suppose there are *n* lines to be processed.  We must loop through them all, so the computational cost is *O(n)* . But we have a loop inside that loop, that invents a variable name that is not allowed to conflict with other variables names, and I (stupidly) am doing that in a second loop that costs *O(n)*.  Therefore, the cost is *O(n^2)$, which is okay for cases like *n=5* that I had in mind when I coded this.  I simply did not envision there being a case with *n=10,000*, which gives cost *O(10^8)*.  I'll think about a faster algorithm for that part of the processing. 

PS. in computing, to say a cost is *O(n)* means that the time is some constant number times *n*.  We ignore the constant factor, which will vary with the actual processing steps and also for the machine.  We focus only on whether it is e.g. *O(n)* (called "linear") or *O(n^2)* ("quadratic") or worse.  For large, *n*, an *O(n^2)* method is terrible.  For example, a bubble sort costs *O(n^2)* but the quick-sort method is *O(n*log(n))*.  Similarly, the FFT is favoured because it lowers the cost in such a way.

**Comment by dankelley** [comment]
I think this is solved in the "develop" branch, commit  2f665346bbf44f3fe864eabe6c847b5d6f36784f 

Perhaps you can check, @clayton33 ?

A sample test code is below.  These files can be read in about 5s each.  Before, they took way longer.  (I don't actually know how slow it was before, since I stopped the process after maybe a minute.)

```R
library(oce)
files <- c("CTD_HUD2013037_042_1_DN.ODF", "CTD_HUD2015030_127_1_DN.ODF")
for (file in files) {
    message(file)
    print(system.time({d <- read.oce(files[1])}))
    summary(d)
}
```

**Comment by clayton33** [comment]
Wow. This looks good :)
